src/adapter/graphql/src/mutations/dataset_mut/adapters_mut/versioned_file_mut.rs

```diff
@@ -393,6 +332,7 @@ pub enum UpdateVersionResult {
     Success(UpdateVersionSuccess),
     CasFailed(UpdateVersionErrorCasFailed),
     InvalidExtraData(UpdateVersionErrorInvalidExtraData),
+    TooLarge(UploadVersionErrorTooLarge),
```

Comment:
Don't think this error belongs in `UpdateVersionResult`. It is returned from `finish_upload_new_version` - and it would be weird to make users expect `TooLarge` errors at this stage as it should've been already rejected at `start_upload` stage.

What I think we need is a separate error for `upload_new_version` - an in-band case that combines `start_upload` and `finish_upload` into one. This one is not meant for production, so it should be OK to change its error type.

We should warn customers though before deploying.

Reply:
As per my other comment, I think we need two limits:
- overall limit for file content size
- limit for in-band upload - to prevent abuse of testing-only API

The latter one is not something we can easily enforce. Yes we can check the length of data in the handler, but by that time **the abuse is already done** - our API got tied up processing absurdly large request.

Perhaps exceeding in-band limits should be handled on the protocol level - by imposing overall limits on the size of the request payload. i.e. so that GQL library will read at most 1MiB from the TCP stream before rejecting the request outright.

Large requests is a general anti-DoS measure which we could skip for now and track in our security review ticket:
https://github.com/kamu-data/planning-internal/issues/62